ConditionalCompilation.md

```diff
@@ -1,28 +1,140 @@
+| (-) poor IDE support | (+) IDE can check all possible code paths |
+
+## Proposal
+
+We propose to leverage the [*attribute* syntax](https://www.w3.org/TR/WGSL/#attributes) to decorate syntax nodes with conditional compilation attributes.
+The compiler is provided with a set of '*compile-time features*' to enable (see below).
+
+### Definitions
+
+ * **Compile-time expression**: A *compile-time expression* is evaluated by the compiler before the rest of the program. It lives in the *compile-time scope*.
+
+ * **Compile-time scope**: The *compile-time scope* this is an independent scope from the [*module scope*](https://www.w3.org/TR/WGSL/#module-scope), meaning it cannot see any declarations from the source code, and its identifers are independent.
+
+ * **Compile-time feature**: A *compile-time feature* is an identifier that evaluates to a boolean. It is set to `true` if the feature is *enabled* during the compilation phase.
+
+ * **Compile-time attribute**: A *compile-time attribute* is parametrized by a *compile-time expression*. It is eliminated after the compilation phase but can affect the syntax node it decorates.
+
+### Location of *Compile-time attributes*
+
+(TODO: this needs more reflexion)
+
+A Compile-time attribute CAN decorate the following syntax nodes:
+ * structure declarations
+ * structure members
+ * function declarations
+ * ?? function formal parameter declarations ?? (is this needed?)
+ * variable declarations and value declarations
+ * all statements, except those disallowed below
+ * directives
+
+Because eliminating the decorated node would lead to invalid code, a Compile-time attribute CANNOT decorate the following syntax nodes:
+ * bodies of function declarations, switch statements, switch clauses, loop statements, for statements, while statements, if/else statements
+ * function return types
+ 
+> Note: The WGSL grammar currently does not allow attributes in front of const value declarations, variable declarations, directives, switch clauses and several statements. We would extend the syntax to allow them.
+> Which exacly we will enable is subject to discussion. Enabling them before all statements is perhaps undesirable.
+
+### Syntax 1: Features only and the `@ifdef`, `@ifndef` attributes
+
+> This is the least complex implementation.
+
+A *compile-time expression* must be a single *compile-time feature*.
+
+Two new attributes are introduced that take a single parameter.
+* The `@ifdef` attribute marks the decorated node for removal if the parameter evaluates to `false`.
+* The `@ifndef` attribute marks the decorated node for removal if the parameter evaluates to `true`.
+
+*Example*
+
+```rs
+@ifdef(my_feature)
+fn f() { ... }
+@ifndef(my_feature)
+fn f() { ... }
+```
+
+### Syntax 2: `@if` attribute with compile-time logical expressions
+
```

Comment:
Syntax 1 and Syntax 2 are alternate, or is the idea to do both?

Given what people are doing, I suspect we need the flexibility of multiple conditions.

Reply:
I think I'll just keep syntax 2. Before writing this I wasn't sure about this so I wrote both. But the idea was to keep one.

Reply:
because at first syntax 2 seemed too complicated, but then I realized that we could just support a simple syntax consisiting of idents, &&, || and !. That can be expanded in thefuture.

Reply:
Sounds good!